proxy/src/transparency/client.rs

```diff
@@ -169,12 +169,32 @@ where
     }
 
     fn call(&mut self, req: Self::Request) -> Self::Future {
+        use http::header::CONTENT_LENGTH;
+
         match self.inner {
             ClientServiceInner::Http1(ref h1) => {
+                // As of hyper 0.11.x, a set body implies a body must be sent.
+                // If there is no content-length header, hyper adds a
+                // transfer-encoding: chunked header, since it has no other way
+                // of knowing if the body is empty or not.
+                //
+                // However, if we received a `Content-Length: 0` body ourselves,
+                // the `body.is_end_stream()` would be true. While its true that
+                // semantically a client request with no body headers and a request
+                // with `Content-Length: 0` have the exact same body length, we
+                // want to preserve the exact intent of the original request, as
+                // we are trying to be a transparent proxy.
+                //
+                // So, if `Content-Length` is set at all, we need to send it. If
+                // we unset the body, hyper assumes the RFC7230 semantics that
+                // it can strip content body headers. Therefore, even if the
+                // body is empty, it should still be passed to hyper so that
+                // the `Content-Length: 0` header is not stripped.
                 let is_body_empty = req.body().is_end_stream();
+                let has_content_length = req.headers().contains_key(CONTENT_LENGTH);
                 let mut req = hyper::Request::from(req.map(BodyStream::new));
-                if is_body_empty {
-                    req.headers_mut().set(hyper::header::ContentLength(0));
+                if is_body_empty && !has_content_length {
```

Comment:
Bigger issue: I have the same question I asked before: Why isn't this check simply `!req.headers().contains_key(CONTENT_LENGTH) && !req.headers().contains_key(TRANSFER_ENCODING)`? If we can use `req.headers().contains_key(TRANSFER_ENCODING)` instead of `req.body().is_end_stream()` then that's much clearer to me. Otherwise, we should add a note about why we can't use a test for the presence of a Transfer-Encoding header here.

(Again, I'm hoping the code will mirror what the RFC says as closely as possible.)